In [20]:
# 1.1 Data Generation
import os
save_path = "/Users/Air/Documentss/LittleSteps-Analysis/data"
os.makedirs(save_path, exist_ok=False)

csv_file = os.path.join(save_path, "visits.csv")
import pandas as pd
import numpy as np 
import random
from datetime import datetime, timedelta

np.random.seed(42)
random.seed(42)

number_patients = 100
number_nurses = 20
number_visits = 600

service_types = [
    "Medication Administration", "Wound Care", "Physical Therapy", "General Check-up",
    "Med Admin", "Wund Care", "Phisycal Therapy", "Check up"
]

visit_locations = ["North", "South", "East", "West",
                   "Nrth", "Suth", "Esst", "Weerst"]

nurse_notes_options = [
    "Patient stable.", 
    "Follow-up urgently required",
    "",
    None, 
    "Wound will be dressed soon",
    "NA",
    "Nil",
    "Urgent attention needed",
    "Schedule revisit in 2 days",
    "Lorem ipsum text $%^&"
]

data = []
for i in range(number_visits):
    visit_id = f"{i+1:04d}"
    patient_id = f"P{random.randint(1, number_patients):03d}"
    nurse_id = f"N{random.randint(1, number_nurses):02d}"
    
    start = datetime(2024, 1, 1) + timedelta(days=random.randint(0, 90),
                                             hours=random.randint(0, 23),
                                             minutes=random.randint(0, 59))
    
    if random.random() < 0.1:
        duration_minutes = random.choice([1, 5, 600, 1440])
    else:
        duration_minutes = random.randint(10, 180)
    
    end = start + timedelta(minutes=duration_minutes)
    
    if random.random() < 0.05:
        end = None
    
    service_type = random.choice(service_types)
    visit_location = random.choice(visit_locations)
    nurse_notes = random.choice(nurse_notes_options)

    if random.random() > 0.1:
        visit_start_time = start.strftime("%Y-%m-%d %H:%M:%S")
    else:
        visit_start_time = start.strftime("%d-%m-%Y %H:%M")

    if end:
        if random.random() > 0.1:
            visit_end_time = end.strftime("%Y-%m-%d %H:%M:%S")
        else:
            visit_end_time = end.strftime("%d/%m/%Y %H:%M")
    else:
        visit_end_time = None

    data.append([
        visit_id, patient_id, nurse_id,
        visit_start_time, visit_end_time,
        service_type, visit_location, nurse_notes
    ])

data.append(data[50])
data.append(data[103])
data.append(data[432])

columns = ["visit_id", "patient_id", "nurse_id",
           "visit_start_time", "visit_end_time",
           "service_type", "visit_location", "nurse_notes"]

df = pd.DataFrame(data, columns=columns)

df.to_csv(csv_file, index=False)

print(f"✅ visits.csv generated at {csv_file} with {len(df)} rows (including duplicates)")


FileExistsError: [Errno 17] File exists: '/Users/Air/Documentss/LittleSteps-Analysis/data'

In [ ]:
# 1.2 Data Loading & Exploration 
df = pd.read_csv("/Users/Air/Documentss/LitteSteps-Analysis/data/visits.csv")

import pandas as pd

display(df.head())

print(df.shape)

display(df.describe())

print(df.isna().sum())

print(df['service_type'].unique())

print(df['visit_location'].unique())

visit_id patient_id nurse_id     visit_start_time       visit_end_time  \
0         1       P082      N04     04-01-2024 23:17  2024-01-05 00:02:00   
1         2       P065      N20  2024-01-04 17:12:00  2024-01-04 19:41:00   
2         3       P055      N11  2024-02-05 04:13:00  2024-02-05 05:49:00   
3         4       P006      N15     09-03-2024 03:59  2024-03-09 06:30:00   
4         5       P038      N03  2024-01-30 03:24:00  2024-01-30 06:16:00   

       service_type visit_location                  nurse_notes  
0        Wound Care           Esst              Patient stable.  
1          Check up           Nrth              Patient stable.  
2  Phisycal Therapy          South                          NaN  
3         Wund Care           West  Follow-up urgently required  
4  Physical Therapy           Suth                          NaN

(603, 8)


visit_id
count  603.000000
mean   299.980100
std    173.483968
min      1.000000
25%    149.500000
50%    300.000000
75%    449.500000
max    600.000000

visit_id              0
patient_id            0
nurse_id              0
visit_start_time      0
visit_end_time       32
service_type          0
visit_location        0
nurse_notes         167
dtype: int64
['Wound Care' 'Check up' 'Phisycal Therapy' 'Wund Care' 'Physical Therapy'
 'General Check-up' 'Med Admin' 'Medication Administration']
['Esst' 'Nrth' 'South' 'West' 'Suth' 'East' 'Weerst' 'North']


1.1: I have succesfully generated a csv file with 603 entries that simulates the nurse scheduling, patients' records and data characteristics.

1.2: From the exploration of the dataset above, I could see that are series of problems ranging like duplicate entries, typos and even wrong formatting. From .unique(), I am able to identify the entries with typo and mistakes, allowing me to clean the data accurately in 1.3.